File: java_refactoring/src/main/java/org/example/solution/FarmCreatorClient.java
package org.example.solution;

import org.example.solution.farmcreator.CattleFarmCreator;
import org.example.solution.farmcreator.CropsFarmCreator;
import org.example.solution.farmcreator.FarmCreator;
import org.example.solution.farmcreator.FishFarmCreator;

public class FarmCreatorClient {

    // use logger instead of system.out


    public static void main(String[] args) { //Shouldnt stay in the static context
        FarmCreatorClient farmCreatorClient = new FarmCreatorClient();
        farmCreatorClient.createFarmsAll();
    }

    public void createFarmsAll(){
        String cattleFarmName = "Farm1";
        FarmCreator cattleFarmCreator = new CattleFarmCreator(cattleFarmName);
        String cattleFarm = this.createFarm(cattleFarmCreator);
        System.out.println("Created farm: " + cattleFarm);

        String fishFarmName = "Farm2";
        FarmCreator fishFarmCreator = new FishFarmCreator(fishFarmName);
        String fishFarm = this.createFarm(fishFarmCreator);
        System.out.println("Created farm: " + fishFarm);

        String cropsFarmName = "Farm3";
        FarmCreator cropsFarmCreator = new CropsFarmCreator(cropsFarmName);
        String cropsFarm = this.createFarm(cropsFarmCreator);
        System.out.println("Created farm: " + cropsFarm);

    }

    private String createFarm(FarmCreator farmCreator){
        farmCreator = this.createLand(farmCreator);
        farmCreator = this.addPlants(farmCreator);
        TaxInfo taxInfo = farmCreator.createTaxes();
        this.payTaxes(taxInfo);
        return farmCreator.getFarm();
    }

    private FarmCreator createLand(FarmCreator farmCreator){
        int numLand = 30;
        System.out.println("Creating number of lands: " + numLand);

        for(int i=0 ; i < numLand; i++) {
            farmCreator.createLand();
        }
        return farmCreator; // preferable to not rely on pass by reference to keep the code more explainable
    }

    private FarmCreator addPlants(FarmCreator farmCreator){
        int numPlants = 20;
        System.out.println("Creating number of plants: " + numPlants);

        for(int i=0 ; i < numPlants; i++) {
            farmCreator.addPlants();
        }
        return farmCreator; // preferable to not rely on pass by reference to keep the code more explainable
    }


    // Moved out of static conext
    private void payTaxes(TaxInfo taxInfo) {
        System.out.println("paying taxes");

        System.out.println("Paying total tax: " + taxInfo.getTotalTax() + " property tax: " + taxInfo.getPropertyTax());
        System.out.println("Paid total tax: " + taxInfo.getTotalTax() + " property tax: " + taxInfo.getPropertyTax());

    }
}

File: typescript_refactoring/example/solution/FarmCreatorClient.ts
import { CattleFarmCreator } from "./CattleFarmCreator";
import { CropsFarmCreator } from "./CropsFarmCreator";
import { FarmCreator } from "./FarmCreator";
import { FishFarmCreator } from "./FishFarmCreator";
import { TaxInfo } from "./TaxInfo";

const LAND_COUNT = 30;
const PLANT_COUNT = 20;

export class FarmCreatorClient {
    public static main(): void {
        const client = new FarmCreatorClient();
        client.createAllFarms();
    }

    public createAllFarms(): void {
        this.createFarm(new CattleFarmCreator("Farm1"));
        this.createFarm(new FishFarmCreator("Farm2"));
        this.createFarm(new CropsFarmCreator("Farm3"));
    }

    private createFarm(farmCreator: FarmCreator): void {
        console.log(`Creating number of lands: ${LAND_COUNT}`);
        for (let i = 0; i < LAND_COUNT; i++) {
            farmCreator.createLand();
        }

        console.log(`Creating number of plants: ${PLANT_COUNT}`);
        for (let i = 0; i < PLANT_COUNT; i++) {
            farmCreator.addPlants();
        }

        this.payTaxes(farmCreator.createTaxes());
        console.log(`Created farm: ${farmCreator.getFarm()}`);
    }

    private payTaxes(taxInfo: TaxInfo): void {
        console.log(`Paying total tax: ${taxInfo.totalTax} property tax: ${taxInfo.propertyTax}`);
        console.log(`Paid total tax: ${taxInfo.totalTax} property tax: ${taxInfo.propertyTax}`);
    }
}
